src/route.rs

```diff
@@ -92,42 +116,160 @@ impl Display for RouteSpec {
             segments,
         }
     }
+
+    fn dots(&self) -> usize {
+        self.segments
+            .iter()
+            .filter(|c| matches!(c, Segment::Dot))
+            .count()
+    }
+
+    #[inline]
+    fn inner_match<'router, 'path>(
+        &'router self,
+        mut path: &'path str,
+        captures: &mut Vec<&'path str>,
+    ) -> Option<&'path str> {
+        let mut peek = self.segments.iter().peekable();
+        while let Some(segment) = peek.next() {
+            path = match &*segment {
+                Segment::Exact(e) => {
+                    if path.starts_with(&**e) {
+                        &path[e.len()..]
+                    } else {
+                        return None;
+                    }
+                }
+
+                Segment::Param(_) => {
+                    if path.is_empty() {
+                        return None;
+                    }
+                    match peek.peek() {
+                        None | Some(Segment::Slash) => {
+                            let capture = path.split('/').next()?;
+                            captures.push(capture);
+                            &path[capture.len()..]
+                        }
+
+                        Some(Segment::Dot) => {
+                            let index = path.find(|c| c == '.' || c == '/')?;
+                            if path.chars().nth(index) == Some('.') {
+                                captures.push(&path[..index]);
+                                &path[index..] // we leave the dot so it can be matched by the Segment::Dot
+                            } else {
+                                return None;
+                            }
+                        }
+                        _ => panic!(
+                            "param must be followed by a dot, a slash, or the end of the route"
+                        ),
+                    }
+                }
+
+                Segment::Wildcard => match peek.peek() {
+                    Some(_) => panic!(concat!(
+                        "wildcard must currently be the terminal segment, ",
+                        "please file an issue if you have a use case for a mid-route *"
+                    )),
+                    None => {
+                        captures.push(path);
+                        ""
+                    }
+                },
+
+                Segment::Slash => match (path.chars().next(), peek.peek()) {
+                    (Some('/'), Some(_)) => &path[1..],
+                    (None, None) => path,
+                    (None, Some(Segment::Wildcard)) => path,
+                    _ => return None,
+                },
+
+                Segment::Dot => match path.chars().next() {
+                    Some('.') => &path[1..],
+                    _ => return None,
+                },
+            }
+        }
+
+        Some(path)
+    }
+
+    /// Returns a vec of captured str slices for this routespec
+    #[inline]
```

Comment:
is this really the right choice here and/or above?

Reply:
Yeah these were both driven by the criterion benches. I don't have enough confidence in my understanding of rustc to add something like this without before/after benchmarks. It helped a statistically significant amount